File: integrations/vellum-assistant/src/store-to-session.ts
/**
 * Store tool calls and assistant responses into the Cognee session cache.
 *
 * Ported from store-to-session.py. Two modes:
 *
 *   Normal (tool call): Write a TraceEntry via storeEntry() with
 *     origin_function, method_params, method_return_value, status.
 *
 *   Stop (--stop flag): Read the pending prompt, pair it with the
 *     assistant's final message, store a QAEntry via storeEntry().
 *
 * Both modes also record in the bridge cache (recordBridgeQA /
 * recordBridgeTrace) for later graph sync.
 *
 * Best-effort: never throws from hooks.
 */

import {
  loadConfig,
  hookLog,
  getSessionKey,
  resolveSessionId,
  sanitizeSessionKey,
  resolveHttpEndpoint,
  isServerReady,
  touchActivity,
  bumpSaveCounter,
  stagePendingPrompt,
  consumePendingPrompt,
  recordBridgeQA,
  recordBridgeTrace,
  type PendingPrompt,
} from "./plugin-common.ts";
import { storeEntry, backendReachable } from "./cognee-client.ts";

// ─── Constants ────────────────────────────────────────────────────────────────

const MAX_PARAMS_BYTES = 4000;
const MAX_RETURN_BYTES = 8000;
const MAX_ASSISTANT_BYTES = 8000;

// ─── Helpers ──────────────────────────────────────────────────────────────────

function truncateStr(value: unknown, cap: number): string {
  if (value === null || value === undefined) return "";
  const text = typeof value === "string" ? value : JSON.stringify(value);
  if (text.length <= cap) return text;
  return text.slice(0, cap - 3) + "...";
}

interface InferredStatus {
  status: string;
  errorMessage: string;
}

function inferStatus(payload: StoreToolPayload): InferredStatus {
  const response = (payload.tool_response ?? payload.tool_output ?? "") as unknown;
  if (typeof response === "object" && response !== null) {
    const resp = response as Record<string, unknown>;
    if (resp.is_error || resp.error) {
      const err = String(resp.error ?? resp.message ?? "Tool reported an error.");
      return { status: "error", errorMessage: truncateStr(err, 500) };
    }
  }
  if (typeof payload.error === "string" && payload.error) {
    return { status: "error", errorMessage: truncateStr(payload.error, 500) };
  }
  return { status: "success", errorMessage: "" };
}

// ─── Types ────────────────────────────────────────────────────────────────────

export interface StoreToolPayload {
  tool_name: string;
  tool_input: unknown;
  tool_output: unknown;
  tool_response?: unknown;
  error?: string;
  conversationId?: string;
}

export interface StoreStopPayload {
  assistant_message: string;
  conversationId?: string;
  turn_id?: string;
}

// ─── Session resolution ───────────────────────────────────────────────────────

function loadSession(conversationId?: string): { sessionId: string; dataset: string } {
  const cfg = loadConfig();
  let sessionKey = getSessionKey();
  if (!sessionKey && conversationId) {
    sessionKey = sanitizeSessionKey(conversationId);
  }
  const sessionId = resolveSessionId(sessionKey, cfg.agentName);
  return { sessionId, dataset: cfg.dataset };
}

// ─── Tool call storage ────────────────────────────────────────────────────────

/**
 * Store a PostToolUse event as a TraceEntry.
 *
 * @param payload Tool call details (name, input, output, status).
 */
export async function storeToolCall(payload: StoreToolPayload): Promise<void> {
  const toolName = payload.tool_name || "unknown";
  const toolInput = payload.tool_input ?? {};
  const toolOutput = payload.tool_output ?? payload.tool_response ?? "";

  // Suppress self-reference: Bash calls mentioning 'cognee' are the plugin
  // talking to itself and would recurse.
  if (toolName === "Bash") {
    let cmd = "";
    if (typeof toolInput === "object" && toolInput !== null) {
      cmd = String((toolInput as Record<string, unknown>).command ?? "");
    }
    if (cmd.includes("cognee")) {
      hookLog("skip_self_cognee_bash", { cmd_prefix: cmd.slice(0, 80) });
      return;
    }
  }

  const { status, errorMessage } = inferStatus(payload);

  // Normalize method_params.
  let params: Record<string, string>;
  if (typeof toolInput === "object" && toolInput !== null && !Array.isArray(toolInput)) {
    params = {};
    for (const [k, v] of Object.entries(toolInput as Record<string, unknown>)) {
      params[k] = truncateStr(v, MAX_PARAMS_BYTES);
    }
  } else {
    params = { value: truncateStr(toolInput, MAX_PARAMS_BYTES) };
  }

  const returnValue = truncateStr(toolOutput, MAX_RETURN_BYTES);

  const { sessionId, dataset } = loadSession(payload.conversationId);
  if (!sessionId) {
    hookLog("no_session_id", { tool: toolName });
    return;
  }

  const { baseUrl, apiKey } = resolveHttpEndpoint();
  if (!apiKey) {
    hookLog("store_tool_no_api_key", { tool: toolName });
    return;
  }

  // Readiness gate: if the server is still warming, buffer into the bridge
  // shadow instead of dropping the trace.
  if (!isServerReady()) {
    if (!(await backendReachable(baseUrl, 1500))) {
      const traceText =
        `${toolName} [${status}]\n` +
        `Params: ${JSON.stringify(params)}\n` +
        `Return: ${returnValue}`;
      recordBridgeTrace(sessionId, dataset, traceText);
      bumpSaveCounter("trace");
      hookLog("store_buffered_warming", { hook: "tool", tool: toolName });
      return;
    }
  }

  const entry: Record<string, unknown> = {
    type: "trace",
    origin_function: toolName,
    status,
    method_params: params,
    method_return_value: returnValue,
    error_message: errorMessage,
    generate_feedback_with_llm: false,
  };

  try {
    const ok = await storeEntry(baseUrl, apiKey, entry, dataset, sessionId);
    if (ok) {
      const traceText =
        `${toolName} [${status}]\n` +
        `Params: ${JSON.stringify(params)}\n` +
        `Return: ${returnValue}`;
      recordBridgeTrace(sessionId, dataset, traceText);
      bumpSaveCounter("trace");
      touchActivity();
      hookLog("trace_stored", { tool: toolName, status });
    } else {
      hookLog("trace_store_noresult", { tool: toolName });
    }
  } catch (err) {
    hookLog("trace_store_error", { tool: toolName, error: String(err).slice(0, 200) });
  }
}

// ─── Stop / assistant message storage ─────────────────────────────────────────

/**
 * Store a Stop-hook payload (final assistant message) as a QAEntry.
 * Pairs the pending prompt (staged by store-user-prompt) with the answer.
 *
 * @param payload Stop details including the final assistant message.
 */
export async function storeAssistantStop(payload: StoreStopPayload): Promise<void> {
  const msg = payload.assistant_message;
  if (!msg || msg === "null") return;

  const truncatedMsg = truncateStr(msg, MAX_ASSISTANT_BYTES);

  const { sessionId, dataset } = loadSession(payload.conversationId);
  if (!sessionId) {
    hookLog("no_session_id", { event: "stop" });
    return;
  }

  const sessionKey = getSessionKey() || sanitizeSessionKey(payload.conversationId ?? "");
  const { baseUrl, apiKey } = resolveHttpEndpoint();
  if (!apiKey) {
    hookLog("store_stop_no_api_key");
    return;
  }

  // Readiness gate.
  if (!isServerReady()) {
    if (!(await backendReachable(baseUrl, 1500))) {
      const pending = consumePendingPrompt(sessionKey);
      recordBridgeQA(sessionId, dataset, pending?.prompt ?? "", truncatedMsg);
      bumpSaveCounter("answer");
      hookLog("store_buffered_warming", { hook: "stop" });
      return;
    }
  }

  // Consume the pending prompt that was staged by store-user-prompt.
  const pending: PendingPrompt | null = consumePendingPrompt(sessionKey);

  const entry: Record<string, unknown> = {
    type: "qa",
    question: pending?.prompt ?? "",
    answer: truncatedMsg,
    context: "",
  };

  try {
    const ok = await storeEntry(baseUrl, apiKey, entry, dataset, sessionId);
    if (ok) {
      recordBridgeQA(sessionId, dataset, pending?.prompt ?? "", truncatedMsg);
      bumpSaveCounter("answer");
      touchActivity();
      hookLog("stop_stored", { chars: truncatedMsg.length });
    }
  } catch (err) {
    hookLog("stop_store_error", { error: String(err).slice(0, 200) });
  }
}

export default { storeToolCall, storeAssistantStop };

File: integrations/vellum-assistant/scripts/store-to-session.py
#!/usr/bin/env python3
"""Store tool calls and assistant responses into the Cognee session cache.

Routes tool calls to the structured ``TraceEntry`` path (new trace-step
shape with origin_function / method_params / method_return_value /
status). Routes the final assistant message on Stop to a ``QAEntry``.

Runs async on the PostToolUse / Stop hooks - fire-and-forget, never
blocks Codex.

Configuration:
    Resolves session state via Cognee HTTP endpoints.
"""

import asyncio
import json
import os
import sys

# Add scripts dir to path for helper imports
sys.path.insert(0, os.path.dirname(__file__))
from _plugin_common import (
    append_http_bridge_entry,
    bump_save_counter,
    bump_turn_counter,
    get_session_key,
    hook_log,
    http_api_ready,
    load_resolved,
    notify,
    persist_session_cache_to_graph_via_http,
    pop_pending_prompt,
    quiet_hook_output,
    remember_entry_via_http,
    resolve_runtime_mode,
    resolve_session_key_from_payload,
    resolve_user,
    server_ready_hint,
    set_session_key,
    touch_activity,
)
from config import (
    ensure_cognee_ready,
    ensure_dataset_ready,
    get_dataset,
    get_session_id,
    load_config,
    persist_session_cache_to_graph,
    sync_graph_context_to_session,
)

# Hard cap per field to avoid ballooning the cache with massive tool outputs.
_MAX_PARAMS_BYTES = 4000
_MAX_RETURN_BYTES = 8000
_MAX_ASSISTANT_BYTES = 8000


async def _fire_improve_background(dataset: str, session_id: str, user, reason: str) -> None:
    """Fire-and-forget session bridge; failures are logged but never raised."""
    try:
        if http_api_ready():
            wrote = persist_session_cache_to_graph_via_http(dataset, session_id)
            hook_log(
                "auto_bridge_fired",
                {"reason": reason, "session": session_id, "via": "http_remember", "wrote": wrote},
            )
            if wrote:
                notify(f"session bridge persisted ({reason})")
            return

        await ensure_dataset_ready(dataset, user)
        wrote = await persist_session_cache_to_graph(dataset, session_id, user)
        graph_result = await sync_graph_context_to_session(dataset, session_id, user)
        hook_log(
            "auto_bridge_fired",
            {
                "reason": reason,
                "session": session_id,
                "wrote": wrote,
                "graph_synced": graph_result.get("synced", 0),
            },
        )
        notify(f"session bridge persisted ({reason})")
    except Exception as exc:
        hook_log("auto_bridge_error", {"reason": reason, "error": str(exc)[:200]})


def _truncate_str(value, cap: int) -> str:
    """Coerce to string and cap at ``cap`` bytes (utf-8), appending ``...`` if truncated."""
    if value is None:
        return ""
    text = value if isinstance(value, str) else json.dumps(value, default=str, ensure_ascii=False)
    encoded = text.encode("utf-8", errors="replace")
    if len(encoded) <= cap:
        return text
    return encoded[: cap - 3].decode("utf-8", errors="ignore") + "..."


def _infer_status(payload: dict) -> tuple[str, str]:
    """Return (status, error_message) from a PostToolUse payload."""
    # Codex and Claude-style payloads may set tool_response.is_error=True on failures; also
    # check for an explicit 'error' key at the top level.
    response = payload.get("tool_response") or payload.get("tool_output") or ""
    if isinstance(response, dict):
        if response.get("is_error") or response.get("error"):
            err = response.get("error") or response.get("message") or "Tool reported an error."
            return "error", _truncate_str(err, 500)
    if isinstance(payload.get("error"), str) and payload["error"]:
        return "error", _truncate_str(payload["error"], 500)
    return "success", ""


def _load_session() -> tuple[str, str, str]:
    """Load session_id, dataset, user_id from resolved cache with fallbacks."""
    resolved = load_resolved()
    session_id = resolved.get("session_id", "")
    dataset = resolved.get("dataset", "")
    user_id = resolved.get("user_id", "")
    if not session_id or not dataset:
        config = load_config()
        session_id = session_id or get_session_id(config)
        dataset = dataset or get_dataset(config)
    return session_id, dataset, user_id


async def _store_tool_call(payload: dict) -> None:
    """Write a PostToolUse event as a TraceEntry."""
    tool_name = payload.get("tool_name", "unknown")
    tool_input = payload.get("tool_input") or {}
    tool_output = payload.get("tool_output") or payload.get("tool_response") or ""

    # Suppress self-reference: any Bash call that mentions 'cognee' is
    # likely the plugin/CLI talking to itself and would recurse.
    if tool_name == "Bash":
        cmd = ""
        if isinstance(tool_input, dict):
            cmd = str(tool_input.get("command", ""))
        if "cognee" in cmd:
            hook_log("skip_self_cognee_bash", {"cmd_prefix": cmd[:80]})
            return

    status, error_message = _infer_status(payload)

    # Normalize method_params: small structured dict is ideal; fall back
    # to a truncated-string dict if we got something non-JSON-safe.
    if isinstance(tool_input, dict):
        params = {}
        for k, v in tool_input.items():
            params[k] = _truncate_str(v, _MAX_PARAMS_BYTES)
    else:
        params = {"value": _truncate_str(tool_input, _MAX_PARAMS_BYTES)}

    return_value = _truncate_str(tool_output, _MAX_RETURN_BYTES)

    session_id, dataset, user_id = _load_session()
    if not session_id:
        hook_log("no_session_id", {"tool": tool_name})
        return

    config = load_config()
    runtime = resolve_runtime_mode()
    use_http = runtime["mode"] == "http"
    hook_log(
        "mode_decision",
        {
            "hook": "store-to-session:tool",
            "mode": runtime["mode"],
            "base_url": runtime.get("base_url", ""),
            "url_source": runtime.get("url_source", ""),
            "key_source": runtime.get("key_source", ""),
            "api_key_present": runtime.get("api_key_present", False),
        },
    )
    if not server_ready_hint(runtime.get("base_url", "")):
        # Server still warming: don't block the tool call and don't lose the
        # trace. Mirror it into the local bridge shadow; the session->graph
        # sync drains it once the server is ready.
        trace_text = (
            f"{tool_name} [{status}]\n"
            f"Params: {json.dumps(params, ensure_ascii=False)}\n"
            f"Return: {return_value}"
        )
        append_http_bridge_entry(dataset, session_id, trace=trace_text)
        bump_save_counter(session_id, "trace")
        hook_log("store_buffered_warming", {"hook": "tool", "tool": tool_name})
        return
    if not use_http:
        await ensure_cognee_ready(config)

    entry = {
        "type": "trace",
        "origin_function": tool_name,
        "status": status,
        "method_params": params,
        "method_return_value": return_value,
        "error_message": error_message,
        # LLM-backed feedback per step is expensive on a busy session —
        # fall back to the deterministic one-liner. Users who want the
        # LLM summary can flip this in a future config.
        "generate_feedback_with_llm": False,
    }

    try:
        if use_http:
            result = remember_entry_via_http(dataset, session_id, entry)
            user = None
        else:
            import cognee
            from cognee.memory import TraceEntry

            user = await resolve_user(user_id)
            result = await cognee.remember(
                TraceEntry(**entry),
                dataset_name=dataset,
                session_id=session_id,
                self_improvement=False,
                user=user,
            )
    except Exception as exc:
        hook_log("trace_store_error", {"tool": tool_name, "error": str(exc)[:200]})
        notify(f"trace store failed ({exc})")
        return

    if result:
        trace_id = (
            result.get("entry_id")
            if isinstance(result, dict)
            else getattr(result, "entry_id", None)
        )
        hook_log(
            "trace_stored",
            {
                "tool": tool_name,
                "status": status,
                "trace_id": trace_id,
            },
        )
        notify(f"trace stored ({tool_name}, {status})")
        if use_http:
            trace_text = (
                f"{tool_name} [{status}]\n"
                f"Params: {json.dumps(params, ensure_ascii=False)}\n"
                f"Return: {return_value}"
            )
            append_http_bridge_entry(
                dataset,
                session_id,
                trace=trace_text,
            )
        bump_save_counter(session_id, "trace")

        touch_activity()
        count, should_improve = bump_turn_counter(session_id)
        if should_improve:
            await _fire_improve_background(dataset, session_id, user, reason=f"turn_{count}")
    else:
        hook_log("trace_store_noresult", {"tool": tool_name})


async def _store_assistant_stop(payload: dict) -> None:
    """Write a Stop-hook payload (final assistant message) as a QAEntry."""
    msg = str(payload.get("assistant_message") or payload.get("last_assistant_message") or "")
    if not msg or msg == "null":
        return

    msg = _truncate_str(msg, _MAX_ASSISTANT_BYTES)

    session_id, dataset, user_id = _load_session()
    if not session_id:
        hook_log("no_session_id", {"event": "stop"})
        return

    config = load_config()
    runtime = resolve_runtime_mode()
    use_http = runtime["mode"] == "http"
    hook_log(
        "mode_decision",
        {
            "hook": "store-to-session:stop",
            "mode": runtime["mode"],
            "base_url": runtime.get("base_url", ""),
            "url_source": runtime.get("url_source", ""),
            "key_source": runtime.get("key_source", ""),
            "api_key_present": runtime.get("api_key_present", False),
        },
    )
    if not server_ready_hint(runtime.get("base_url", "")):
        # Server still warming: buffer the prompt+answer into the local bridge
        # shadow instead of dropping it; the session->graph sync drains it once
        # the server is ready.
        pending = pop_pending_prompt(session_id, turn_id=str(payload.get("turn_id") or ""))
        append_http_bridge_entry(
            dataset,
            session_id,
            question=pending.get("prompt", ""),
            answer=msg,
        )
        bump_save_counter(session_id, "answer")
        hook_log("store_buffered_warming", {"hook": "stop"})
        return
    if not use_http:
        await ensure_cognee_ready(config)

    pending = pop_pending_prompt(session_id, turn_id=str(payload.get("turn_id") or ""))

    # Codex intentionally differs from Claude here: store one paired
    # prompt/answer row so Cognee's filesystem session cache does not get
    # separate question-only and answer-only QA entries for the same turn.
    entry = {
        "type": "qa",
        "question": pending.get("prompt", ""),
        "answer": msg,
        "context": pending.get("context", ""),
    }

    try:
        if use_http:
            result = remember_entry_via_http(dataset, session_id, entry)
            user = None
        else:
            import cognee
            from cognee.memory import QAEntry

            user = await resolve_user(user_id)
            result = await cognee.remember(
                QAEntry(**entry),
                dataset_name=dataset,
                session_id=session_id,
                self_improvement=False,
                user=user,
            )
    except Exception as exc:
        hook_log("stop_store_error", {"error": str(exc)[:200]})
        notify(f"stop store failed ({exc})")
        return

    if result:
        if use_http:
            append_http_bridge_entry(
                dataset,
                session_id,
                question=pending.get("prompt", ""),
                answer=msg,
            )
        qa_id = (
            result.get("entry_id")
            if isinstance(result, dict)
            else getattr(result, "entry_id", None)
        )
        hook_log("stop_stored", {"chars": len(msg), "qa_id": qa_id})
        notify(f"assistant message stored ({len(msg)} chars)")
        bump_save_counter(session_id, "answer")

        touch_activity()
        count, should_improve = bump_turn_counter(session_id)
        if should_improve:
            await _fire_improve_background(dataset, session_id, user, reason=f"turn_{count}")


def main():
    payload_raw = sys.stdin.read()
    if not payload_raw.strip():
        return

    try:
        payload = json.loads(payload_raw)
    except json.JSONDecodeError:
        hook_log("invalid_payload_json")
        return

    session_key_candidate, session_key_source = resolve_session_key_from_payload(payload)
    if session_key_candidate:
        set_session_key(session_key_candidate)
    hook_log("store_session_key", {"source": session_key_source, "value": session_key_candidate})
    if not get_session_key():
        hook_log("store_missing_session_key")
        return

    is_stop = "--stop" in sys.argv
    try:
        with quiet_hook_output("store-to-session"):
            if is_stop:
                asyncio.run(_store_assistant_stop(payload))
            else:
                asyncio.run(_store_tool_call(payload))
    except Exception as exc:
        hook_log("run_exception", {"stop": is_stop, "error": str(exc)[:200]})


if __name__ == "__main__":
    main()
